iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/command/CommandMatcher.java

```diff
@@ -0,0 +1,287 @@
+      if (is(Status.ABORT)) {
+        // could also check for FOUND_NO_PREDICATE_MATCH, but renders the boolean anythingMatched useless which is added for better readability of
+        // the control flow
+        return;
+      }
+
+    }
+
+    // calculate success or failure due to size of sets
+    if (fulfilledOptions.size() != expectedMatches) {
+      // not all expectedMatches are fulfilled so overall the MultipleUnorderedOptionsPredicate hasn't matched
+      setStatus(Status.ABORT);
+      return;
+    }
+    setStatus(Status.CONTINUE);
+  }
+
+  public boolean matchExpectedOrAnyOption(Set<OptionPredicate> fulfilledOptions, List<OptionPredicate> workingSet, OptionPredicate anyOptionExceptExpectedPredicate) {
+    boolean anythingMatched = false;
+    // used to gauge if there happened to be a match in this iteration of the while-loop
+    int previousNumberOfCommandArguments = argumentsToReport.size();
+
+    Iterator<OptionPredicate> iterator = workingSet.iterator();
+    // this loop tries to match (only) one option of the workingSet to the first resolution of the argument Stack
+    while (iterator.hasNext() && !anythingMatched) {
+      OptionPredicate option = iterator.next();
+
+      // saving this resolution to add it back to the stack if there was no match on the flag
+      ArgumentResolution argumentResolution = argumentStack.peekFirst();
+
+      matchPredicate(option);
+
+      if (is(Status.FOUND_NO_PREDICATE_MATCH)) {
+        // if the flag of an option doesn't match, the argumentResolution has to be readded to the stack for the next possible option
+        argumentStack.addFirst(argumentResolution);
+      }
+
+      if (is(Status.ABORT)) {
+        return false;
+      }
+
+      // found a matching option
+      if (previousNumberOfCommandArguments != argumentsToReport.size()) {
+        iterator.remove();
+        fulfilledOptions.add(option);
+        anythingMatched = true;
+      }
+    }
+
+    // no match in workingSet found, if desired matches any other flags
+    if (!anythingMatched && ((MultipleUnorderedOptionsPredicate) currentPredicate).isShouldSupportAnyMatch()) {
+      matchPredicate(anyOptionExceptExpectedPredicate);
+
+      // found a matching option
+      if (previousNumberOfCommandArguments != argumentsToReport.size()) {
+        anythingMatched = true;
+      }
+    }
+
+    return anythingMatched;
+  }
+
+  public boolean is(Status... statusArray) {
+    for (Status specificStatus : statusArray) {
+      if (this.status.equals(specificStatus)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public void setStatus(Status status) {
+    this.status = status;
+  }
+
+  private ArgumentResolution pollFirstArgumentResolution() {
+    return argumentStack.pollFirst();
+  }
+
+  private ArgumentResolution peekFirstArgumentResolution() {
+    return argumentStack.peekFirst();
+  }
```

Comment:
Not totally convinced that those method are necessary ^^ I don't have a strong opinion on this, feel free to keep them if you think they bring clarity.

Reply:
This was my suggestion, as it might be hard to understand what peekFirst means for argument handling. And I agree that this naming does not bring any big advantages as it duplicates the `argumentStack.peekFirst` wording.

Reply:
Yeah, looking over it again, I'm also not convinced about those methods. Because we are using the argumentStack now as an argument in the CommandMatcher class, we can drop the extra method call. Or are there other suggestions?